src/solutions/optimization_solutions.jl

```diff
@@ -17,28 +17,43 @@ Representation of the solution to an nonlinear optimization defined by an Optimi
   callback (`sol.retcode === :Terminated`), or whether it exited due to an error. For more
   details, see the return code section of the DifferentialEquations.jl documentation.
 """
-struct OptimizationSolution{T, N, uType, P, A, Tf, O} <: AbstractOptimizationSolution{T, N}
+struct OptimizationSolution{T, N, uType, C, A, Tf, O} <: AbstractOptimizationSolution{T, N}
     u::uType # minimizer
-    prob::P # optimization problem
+    cache::C # optimization cache
     alg::A # algorithm
     minimum::Tf
     retcode::ReturnCode.T
     original::O # original output of the optimizer
 end
 
-function build_solution(prob::AbstractOptimizationProblem,
-                        alg, u, minimum;
-                        retcode = ReturnCode.Default,
-                        original = nothing,
-                        kwargs...)
+function Base.getproperty(sol::OptimizationSolution, sym::Symbol)
+    if sym == :prob
+        Base.depwarn("`sol.prob` is deprecated. Use `sol.cache` instead.", "sol.prob")
+        return getfield(sol, :cache)
+    end
+    return getfield(sol, sym)
+end
+
+function build_optimization_solution(cache,
+                                     alg, u, minimum;
+                                     retcode = ReturnCode.Default,
+                                     original = nothing,
+                                     kwargs...)
     T = eltype(eltype(u))
     N = ndims(u)
 
-    OptimizationSolution{T, N, typeof(u), typeof(prob), typeof(alg),
-                         typeof(minimum), typeof(original)}(u, prob, alg, minimum, retcode,
+    OptimizationSolution{T, N, typeof(u), typeof(cache), typeof(alg),
+                         typeof(minimum), typeof(original)}(u, cache, alg, minimum, retcode,
                                                             original)
 end
 
+function build_solution(prob::AbstractOptimizationProblem, args...; kwargs...)
```

Comment:
why not just be adding a dispatch here like all of the other problems? Why does it need a separate function?

Reply:
I will add an `AbstractOptimizationCache` and dispatch on it?